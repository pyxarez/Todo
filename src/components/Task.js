import React from 'react';

import Button from './Button';
import './Task.res/style.css';

Task.propTypes = {
  isDone: React.PropTypes.bool.isRequired,
  title: React.PropTypes.string,
  handleDoneTaskClick: React.PropTypes.func.isRequired,
  handleEditTaskClick: React.PropTypes.func.isRequired
}

export default function Task({ isDone, title, handleDoneTaskClick, handleEditTaskClick }) {
  return (
    <div className="my-task-component">
      <input onChange={handleDoneTaskClick} checked={isDone} type="checkbox"/>
      <h2>{title}</h2>
      <Button onClick={handleEditTaskClick} type="edit" />
    </div>
  );
}