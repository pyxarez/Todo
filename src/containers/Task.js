import React, { Component, PropTypes } from 'react';
import { browserHistory } from 'react-router';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { checkDone } from '../store/actions/TaskListActions';
import { getProgress } from '../store/actions/ProgressActions';

import Task from '../components/Task';

export class TaskContainer extends Component {
  static propTypes = {
    categoryId: PropTypes.number.isRequired,
    task: PropTypes.shape({
      id: PropTypes.node.isRequired,
      title: PropTypes.string.isRequired,
      isDone: PropTypes.bool.isRequired,
      description: React.PropTypes.string.isRequired
    }),
    checkDone: PropTypes.func.isRequired,
    getProgress: React.PropTypes.func.isRequired,
  }

  handleEditTaskClick = () => {
    const categoryId = this.props.categoryId;
    const {
      id: taskId,
      title
    } = this.props.task;

    const path = `/edit/${categoryId}/${taskId}/${title}`;
    browserHistory.push(path);
  }

  handleDoneTaskClick = () => {
    const {
      checkDone,
      getProgress,
      categoryId
    } = this.props;
    const { id } = this.props.task;

    return checkDone(categoryId, id)
      .then(() => { getProgress(); });
  }

  render() {
    const {
      isDone,
      title
    } = this.props.task;

    return (
      <Task
        title={title}
        isDone={isDone}
        handleDoneTaskClick={this.handleDoneTaskClick}
        handleEditTaskClick={this.handleEditTaskClick}/>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    checkDone: bindActionCreators(checkDone, dispatch),
    getProgress: bindActionCreators(getProgress, dispatch)
  }
}

export default connect(null, mapDispatchToProps)(TaskContainer);