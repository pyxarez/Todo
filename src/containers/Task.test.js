import React from 'react';
import { shallow, mount } from 'enzyme';

import { TaskContainer } from './Task';
import Task from '../components/Task';

jest.mock('react-router', () => {
  return {
    browserHistory: {
      push: jest.fn()
    }
  };
});
import { browserHistory } from 'react-router';

const setup = () => {

  const props = {
    categoryId: 0,
    task: {
      id: 0,
      title: 'New Title',
      isDone: false,
      description: '1 2 34 5 6 7'
    },
    checkDone: jest.fn(() => new Promise((resolve, reject) => {
      resolve('Resolved');
    })),
    getProgress: jest.fn()
  }
  const wrapper = shallow(<TaskContainer {...props}/>);

  return {
    wrapper,
    props
  }
};

describe('Containers :: TaskContainer', () => {
  it('should render self and subcomponents', () => {
    const { wrapper } = setup();
    
    expect(wrapper).toMatchSnapshot();
  });

  it('should call getProgress and checkDone', async () => {
    const { props, wrapper } = setup();

    try {
      await wrapper.find(Task).prop('handleDoneTaskClick')();
      expect(props.getProgress).toBeCalled();
      expect(props.checkDone)
        .toBeCalledWith(props.categoryId, props.task.id);
    } catch(e) {
      throw e;
    }
  });

  it('should call browserHistory.push()', () => {
    const { wrapper, props } = setup();

    wrapper.find(Task).prop('handleEditTaskClick')();
    expect(browserHistory.push)
      .toBeCalledWith(`/edit/${props.categoryId}/${props.task.id}/${props.task.title}`);
  });
});