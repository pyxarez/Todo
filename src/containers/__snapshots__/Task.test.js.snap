exports[`Containers :: TaskContainer should render self and subcomponents 1`] = `
<Task
  handleDoneTaskClick={[Function]}
  handleEditTaskClick={[Function]}
  isDone={false}
  title="New Title" />
`;
